### DISCUSSION TITLE
Is it possible to auto-highlight more than one element in a performant way?

### DISCUSSION THREAD

**Discussion by ssaornil** [visgl/deck.gl#5791] [Q&A] [answered]
I am using the PathLayer to display several geojson lineString elements. I want to highlight the hovered element and the last clicked element with the same color.

Since `autoHighlight` is only available for the hovered element I tried using `getColor` to do so but it takes way too long because it is recoloring every single element.

Any idea on how to accomplish the same effect in a more performant way?

**Comment by padawannn** [comment, marked as answer]
Are you using [update triggers](https://deck.gl/docs/developer-guide/performance#use-updatetriggers) with getColor? This way it should work fast

**Reply by ssaornil** [reply]
Yes we are, but if I am not mistaken the update trigger is just a value for each accessor. AFAIK it does not have a different value for each item in the dataset.

So when we trigger it, it calls the `getColor` accesor for each individual item in the dataset. Our accessor is fast but we have many elements ~20K and that makes the re-render quite slow.

**Reply by padawannn** [reply]
If I understand well your use case, `getColor` should return the same color for all elements except the clicked element so `getColor` should only be executed when you click on an element but not when you hovered it.
To change the color of the hovered element use `autoHighlight` and `highlightColor` 

I think it works slow for you because you are using `getColor` to change the color of the hovered and clicked element, I'm right? With the strategy that I mention in the previous paragraph (use `getColor` for clicked and `autoHighlight` for hovered), it should work fast

**Reply by padawannn** [reply]
Another strategy could be to use two layers, the first one would have all the elements and `autoHighlight` property for hovered elements. The second layer will only have the clicked item so you would update the data of the second layer every time you click on the first one.

In any case, I think that what I wrote in the previous comment should work.

**Reply by Pessimistress** [reply]
20k is not a large number, but if you are concerned about perf, you can try [dataDiff](https://deck.gl/docs/api-reference/core/layer#_datadiff)

**Reply by ssaornil** [reply]
@Pessimistress we can clearly see a considerable delay between the appearance of our popover and the coloring of the object and when we checked we saw it was caused by the `getColor` accessor. Our input data is not changing so if I am not mistaken `dataDiff` is not being invoked, we are relying on the `updateTriggers`.

@padawannn you got it completely right, we are using `getColor` to color the highlighted and the clicked element. We'll try to use one of your suggested approaches. Thank you a lot!

**Reply by Pessimistress** [reply]
That shouldn't be the case. If you could share your `getColor` function we can make further optimization suggestions.

If your data prop is not shallowly changed you can use `dataComparator` to trigger a data update.

**Reply by Pessimistress** [reply]
Also: not sure what you are setting your `updateTriggers` to, if you use `autoHighlight` for hovered object and only trigger a color update on click, there shouldn't be too much delay.
